### PR TITLE
Verify toolchain components at program invocation

### PR DESCRIPTION
**Pull request by bstefanuk**
Currently, there is code spread around TensileLt for verifying the selected toolchain, such as checking if a requested compiler is supported, finding the path to the compiler, checking if the offload bundler can be found, and verifying it is executable, among others. Since the toolchain is fixed throughout the execution of Tensile, TensileCreateLibrary, and other programs in the TensileLt project, moving shared toolchain validation logic to a module that is called once in the setup phase of each program improve the maintainability and clarity of the code, but more importantly eliminates occurrences whereby a program crashes several minutes (or more) into its execution because of an improper toolchain component.

**Outcomes**
- *hipcc* is no longer a supported compiler.
- All toolchain verification code is moved to the setup phase of each of TensileLt's programs.
- A new module *Tensile.Utilities.Toolchain* is added to collect shared toolchain verification logic.
- The parameters CxxCompiler, CCompiler, ClangOffloadBundlerPath, and AssemblerPath are no longer globally accessible.
- Setting the path to *clang-offload-bundler* now follows the same CLI pattern as the compiler/assembler, thus the associated environment variable has been removed.

**Principles**
In an effort to improve coding standards for future developments to hipBLASLt, the following principles were followed when implementing this PR.

1. Prefer non-global variables: toolchain components are not globally accessible.
    - Effect: Code that require a toolchain component must accept it as a function or class parameter.
2. Toolchain components must be provided at program invocation, either explicitly or implicitly via defaults.
3. The program will not make assumptions about the toolchain component based on invalid input, e.g., convert *hipcc* input to *amdclang++* under the hood.
4. Environment variables can no longer be used to set toolchain components, only to set search paths
5. Toolchain components will be preferentially searched in /opt/rocm/bin followed by /opt/rocm/lib/llvm/bin followed by the existing PATH variable.
    - Effect: If the user wishes to use a specific non-standard toolchain component, the absolute path must be provided.

**Testing**
- All *tox* tests pass—Environment: Ubuntu 22.04 Docker with ROCm 6.3.0, Python 3.10.12
- Functional testing via routine invocations for *Tensile* and *TensileCreateLibrary* using toolchain defaults.
- [x] TensileCreateLibrary, gfx1101, lazy lib loading, Linux
- [x] TensileCreateLibrary, gfx1101, lazy lib loading, Windows
- [x] Tensile client, Linux
- [x] Tensile client, Windows

### CODE REVIEW

**Comment by bstefanuk** [comment]
*Draft* status will be removed after Windows support is confirmed.

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 1]
```diff
@@ -244,93 +230,83 @@ def buildSourceCodeObjectFile(CxxCompiler, outputPath, kernelFile):
...
+    # needs to be fixed when Maneesh's change is made available
+    hipFlags = ["-D__HIP_HCC_COMPAT_MODE__=1"]
+    hipFlags += ["--cuda-device-only", "-x", "hip", "-O3"]
+    hipFlags += ['-I', outputPath]
+    hipFlags += ["-Xoffload-linker", "--build-id=%s"%globalParameters["BuildIdKind"]]
+    hipFlags += ['-std=c++17']
+    if globalParameters["AsanBuild"]:
+      hipFlags += ["-fsanitize=address", "-shared-libasan", "-fuse-ld=lld"]
+    if globalParameters["SaveTemps"]:
+      hipFlags += ['--save-temps']
```
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py]: If we are getting rid of globals should we remove these? Not sure what `SaveTemps` is and if it is used. 

I think we should consider lifting the details of setting flags that do not change out of this function and set it once.

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 2]
```diff
@@ -244,93 +230,83 @@ def buildSourceCodeObjectFile(CxxCompiler, outputPath, kernelFile):
...
 
-      hipFlags += ['-I', outputPath]
-      hipFlags += ["-Xoffload-linker", "--build-id=%s"%globalParameters["BuildIdKind"]]
-      hipFlags += ['-std=c++17']
-      if globalParameters["AsanBuild"]:
-        hipFlags += ["-fsanitize=address", "-shared-libasan", "-fuse-ld=lld"]
-      if globalParameters["SaveTemps"]:
-        hipFlags += ['--save-temps']
+    launcher = shlex.split(os.environ.get('Tensile_CXX_COMPILER_LAUNCHER', ''))
```
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py]: Can we do this once when we validate the toolchain?

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 3]
```diff
@@ -244,93 +230,83 @@ def buildSourceCodeObjectFile(CxxCompiler, outputPath, kernelFile):
...
-          coFilenames.append(os.path.split(outfile)[1])
-          #bundlerArgs = [globalParameters["ClangOffloadBundlerPath"], "-type=o", "-targets=hip-amdgcn-amd-amdhsa--%s" % cmdlineArchs[i], "-inputs=%s" % infile, "-outputs=%s" % outfile, "-unbundle"]
-          bundlerArgs = [globalParameters["ClangOffloadBundlerPath"], "-type=o", "-targets=hip-amdgcn-amd-amdhsa--%s" % cmdlineArchs[i],
-                         "%s=%s" % (inflag, infile), "%s=%s" % (outflag, outfile), "-unbundle"]
-          if globalParameters["PrintCodeCommands"]:
-            print(' '.join(bundlerArgs))
-          subprocess.check_call(bundlerArgs)
+    if os.name == "nt":
+      hipFlags += ['-fms-extensions', '-fms-compatibility', '-fPIC', '-Wno-deprecated-declarations']
+      compileArgs = launcher + [which(CxxCompiler)] + hipFlags + archFlags + [kernelFile, '-c', '-o', os.path.join(buildPath, objectFilename)]
```
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py]: same thing here - can we separate out the things that only need to be set once.

i.e. 

```
launcher + [which(CxxCompiler)] + hipFlags + archFlags + [kernelFile, '-c', '-o']
```
and pass this in.

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 4]
```diff
@@ -244,93 +230,83 @@ def buildSourceCodeObjectFile(CxxCompiler, outputPath, kernelFile):
...
+    # get hipcc version due to compatiblity reasons
+    hipccver = globalParameters['HipClangVersion'].split(".")
+    hipccMaj = int(hipccver[0])
+    hipccMin = int(hipccver[1])
+    # for hipclang 5.2 and above, clang offload bundler changes the way input/output files are specified
+    inflag = "-inputs"
+    outflag = "-outputs"
+    if (hipccMaj == 5 and hipccMin >= 2) or hipccMaj >= 6:
+      inflag = "-input"
+      outflag = "-output"
```
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py]: seems like we could do most of this once when we validate our toolchain

File: tensilelite/Tensile/TensileCreateLibrary.py [hunk 5]
```diff
@@ -1158,79 +1119,6 @@ def libraryIter(lib: MasterSolutionLibrary):
 
   return solutions, masterLibraries, fullMasterLibrary
 
-################################################################################
```
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py]: Did you remove this or did you merge one of my branches in?

File: tensilelite/Tensile/Utilities/Toolchain.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    Args:
+        executable: The toolchain component to check the version of.
+        versionFlag: The flag to pass to the executable to get the version.
+    """
+    try:
+        output = run([executable, versionFlag], stdout=PIPE, stderr=PIPE, check=True).stdout.decode().strip()
+        match = re.search(regex, output, re.IGNORECASE)
+        return match.group(1) if match else "<unknown>"
+    except Exception as e:
+        raise RuntimeError(f"Failed to get version of {executable}: {e}")
```
Comment (davidd-amd) [inline tensilelite/Tensile/Utilities/Toolchain.py]: need line at end of file

File: tensilelite/Tensile/TensileCreateLibrary.py (hunk 5 above)
Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: I removed these, but I know this change is shadowed in another PR. Ideally, we merge the other PR first.

File: tensilelite/Tensile/Utilities/Toolchain.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        raise RuntimeError(f"Failed to get version when calling {args}: {e}")
+
+
+
+
+
+
+
+
+
```
Comment (bstefanuk) [inline tensilelite/Tensile/Utilities/Toolchain.py]:
```suggestion
```

File: tensilelite/Tensile/SolutionLibrary.py [hunk 8]
```diff
@@ -459,14 +460,15 @@ def selection(d, problemType, solutions, library, placeholderName):
...
+                                                        cxxCompiler,
                                                         solutionClass,
                                                         libraryOrder[placeholderIndex:],
                                                         placeholderName)
             libraryOrder = libraryOrder[0:placeholderIndex]
             origSolutions = []
 
         problemType = Contractions.ProblemType.FromOriginalState(origData["ProblemType"])
-        allSolutions = [solutionClass.FromSolutionStruct(s) for s in origSolutions]
+        allSolutions = [solutionClass.FromSolutionStruct(s, cxxCompiler) for s in origSolutions]
```
Comment (davidd-amd) [inline tensilelite/Tensile/SolutionLibrary.py:471]: Why does the solution need to know about the compiler?

File: tensilelite/Tensile/SolutionStructs.py [hunk 9]
```diff
@@ -1233,7 +1234,7 @@ def initActivationFunctionObjects(self):
       state["ProblemType"]["GroupedGemm"] = False
       state["KernelLanguage"] = "Source"
       state["Kernel"] = {"WavefrontSize": self["WavefrontSize"], "ISA": tuple(self["ISA"])}
-      self.activationFunctionObjects.append(KernelWriterActivationFunction(state))
+      self.activationFunctionObjects.append(KernelWriterActivationFunction(state, self.cxxCompiler))
```
Comment (davidd-amd) [inline tensilelite/Tensile/SolutionStructs.py:1237]: ouch this is why we need to move compile commands out of the Kernel Helpers

Comment (bstefanuk) [reply tensilelite/Tensile/SolutionStructs.py:1237]: Spot on!

**Review by ellosel — DISMISSED** [review]
This looks good to me. We now verify the selected toolchain once at the beginning of the mains that require a toolchain and pass the verified toolchain components to the functions that require them.

File: tensilelite/Tensile/TensileCreateLibrary.py (hunk 1 above)
Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: Out of scope. This behaviour will be added in a subsequent PR.

File: tensilelite/Tensile/TensileCreateLibrary.py (hunk 2 above)
Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: Out of scope. This behaviour will be added in a subsequent PR.

File: tensilelite/Tensile/TensileCreateLibrary.py (hunk 3 above)
Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: Out of scope.

File: tensilelite/Tensile/TensileCreateLibrary.py (hunk 4 above)
Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: Great idea, this code will also be affected by #1374, after which the toolchain consolidation work can be concluded in a final PR.

**Review by KKyang — APPROVED** [review]

**Review by ellosel — APPROVED** [review]

**Pull request merged**